### PR TITLE
llama : minimize size used for state save/load

### PR DESCRIPTION
**Pull request by dfriehs**
#### Context

I am playing with creating a chat application for Android on which I would like to save context state on each message written to allow for chat continuation and message regeneration. As such, small state serialization and robust state deserialization are of great interest.

#### Overview

With the new `llama_batch` API state saving and loading is currently broken as the capacity of `llama_context->logits` depends on the largest batch size ever passed to `llama_decode` (see #4430). To fix this issue, this PR suggests reserving the maximum capacity reachable during context initialization.

To reduce the increase in storage required to now serialize all of the logits buffer, save and load functions are changed to only save currently valid logits. The check against capacity is kept, but as an upper bound instead of requiring it to match serialized state. This enables "compacting" the state by passing a batch of one token as the last batch before saving.

As I saw no reason not to, RNG serialization is also changed to only write the required amount of space instead of a full block of `LLAMA_MAX_RNG_STATE`, although the upper limit is kept.

---

As this changes the session file, this is a **breaking** change. I welcome any kind of feedback or request for changes.

### CODE REVIEW

**Comment by ggerganov** [comment]
Thanks - this looks like a nice improvement. Please not that we will wait for #4766 to get merged first because it is with higher priority and touches the same code. After merging it, this PR can be updated and we can merge it as well

> This enables "compacting" the state by passing a batch of one token as the last batch before saving.

I didn't quite get that. If this requires a specific call with batch of 1 token before `llama_copy_state_data` it should be documented and asserted. Ideally, we would like to not have this constraint, so we should think about workarounds

**Comment by dfriehs** [comment]
Glad to hear!

> I didn't quite get that. If this requires a specific call with batch of 1 token before llama_copy_state_data it should be documented and asserted. Ideally, we would like to not have this constraint, so we should think about workarounds

There is no constraint here, as this is just a side effect of the new batch API: If a batch of one token is passed `context->logits` will be resized to `1 * n_vocab`; which means the save will be the smallest it can be.

**Comment by slaren** [comment]
#4766 only changes the KV state save/load, and this seems unchanged here, so I don't expect any conflicts if this is merged first.
